packages/perseus/src/widgets/interactive-graphs/interactive-graph-reducer.ts

```diff
@@ -117,24 +127,27 @@ function clamp(value: number, min: number, max: number) {
     return value;
 }
 
-function validSegments(segments: Segment[]): boolean {
+function validSegments(segments: readonly CollinearTuple[]): boolean {
     return segments.every(([start, end]) => !kvector.equal(start, end));
 }
 
-function updateAtIndex<A extends readonly any[]>(
-    array: A,
-    index: number,
-    update: (elem: A[number]) => A[number],
-): A {
-    return setAtIndex(array, index, update(array[index]));
+function updateAtIndex<T>(args: {
+    array?: readonly T[];
+    index: number;
+    update: (elem: T) => T;
+}): readonly T[] {
+    const {array = [], index, update} = args;
+    const newValue = update(array[index]);
+    return setAtIndex({array, index, newValue});
 }
 
-function setAtIndex<A extends readonly any[]>(
-    array: A,
-    index: number,
-    newValue: A[number],
-): A {
-    const copy = [...array] as any;
+function setAtIndex<T, A extends readonly T[]>(args: {
+    array?: A;
+    index: number;
+    newValue: A[number];
+}): A {
+    const {array, index, newValue} = args;
+    const copy: T[] = [...(array || [])];
```

Comment:
**suggestion**: Instead of dealign with the default array value on line 150, would this be a bit easier to see?

```suggestion
function setAtIndex<T>({
    array = [], 
    index, 
    newValue
}: {
    array?: ReadonlyArray<T>;
    index: number;
    newValue: T;
}): ReadonlyArray<T> {
    const copy: T[] = [...array];
    copy[index] = newValue;
    return copy;
}
```

Reply:
@jeremywiebe I like adding the default array, that would be cleaner!